File: src/identity.js
/**
 * OCTO Protocol — Identity Module
 * 
 * BIP-39 Mnemonic (24 words) → ed25519 Keypair → Tor v3 .onion Address
 * 
 * The pipeline:
 *   24 words → PBKDF2 → 512-bit seed → first 32 bytes → ed25519 seed
 *   ed25519 seed → SHA-512 expand → private scalar + righthalf (Tor format)
 *   ed25519 seed → public key → SHA3-256 checksum → base32 → .onion address
 */

const bip39 = require('bip39');
const nacl = require('tweetnacl');
const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

// ─── Base32 Encoding (RFC 4648, lowercase) ──────────────────────────

const BASE32_ALPHABET = 'abcdefghijklmnopqrstuvwxyz234567';

function base32Encode(buffer) {
    let bits = '';
    for (const byte of buffer) {
        bits += byte.toString(2).padStart(8, '0');
    }
    let result = '';
    for (let i = 0; i < bits.length; i += 5) {
        const chunk = bits.substring(i, i + 5).padEnd(5, '0');
        result += BASE32_ALPHABET[parseInt(chunk, 2)];
    }
    return result;
}

// ─── Mnemonic Generation ────────────────────────────────────────────

/**
 * Generate a new 24-word BIP-39 mnemonic (256 bits of entropy)
 */
function generateMnemonic() {
    return bip39.generateMnemonic(256);
}

/**
 * Validate a BIP-39 mnemonic phrase
 */
function validateMnemonic(mnemonic) {
    return bip39.validateMnemonic(mnemonic);
}

// ─── Key Derivation ─────────────────────────────────────────────────

/**
 * Convert mnemonic to 64-byte seed via PBKDF2
 */
function mnemonicToSeed(mnemonic) {
    return bip39.mnemonicToSeedSync(mnemonic);
}

/**
 * Derive ed25519 keypair from a BIP-39 seed
 * Takes the first 32 bytes of the 64-byte seed as the ed25519 seed
 */
function seedToKeyPair(seed) {
    const ed25519Seed = Buffer.from(seed.slice(0, 32));
    const keyPair = nacl.sign.keyPair.fromSeed(ed25519Seed);
    return {
        publicKey: Buffer.from(keyPair.publicKey),
        secretKey: Buffer.from(keyPair.secretKey),
        seed: ed25519Seed
    };
}

/**
 * Expand ed25519 seed into Tor's secret key format
 * SHA-512(seed) → clamp first 32 bytes (scalar) + keep last 32 bytes (righthalf)
 */
function expandSecretKey(ed25519Seed) {
    const h = crypto.createHash('sha512').update(ed25519Seed).digest();
    // Clamp the scalar (first 32 bytes)
    h[0] &= 248;    // Clear bits 0, 1, 2
    h[31] &= 127;   // Clear bit 255
    h[31] |= 64;    // Set bit 254
    return {
        scalar: h.slice(0, 32),
        righthalf: h.slice(32, 64)
    };
}

// ─── .onion Address Computation ─────────────────────────────────────

/**
 * Compute Tor v3 .onion address from ed25519 public key
 * 
 * Formula: base32(PUBKEY || CHECKSUM || VERSION) + ".onion"
 * CHECKSUM = SHA3-256(".onion checksum" || PUBKEY || VERSION)[:2]
 * VERSION = 0x03
 */
function publicKeyToOnion(publicKey) {
    const version = Buffer.from([0x03]);
    const prefix = Buffer.from('.onion checksum');

    // Compute checksum: first 2 bytes of SHA3-256(prefix + pubkey + version)
    const checksum = crypto.createHash('sha3-256')
        .update(prefix)
        .update(publicKey)
        .update(version)
        .digest()
        .slice(0, 2);

    // Concatenate: pubkey(32) + checksum(2) + version(1) = 35 bytes
    const addressBytes = Buffer.concat([publicKey, checksum, version]);

    // Base32 encode → 56-character .onion address
    return base32Encode(addressBytes) + '.onion';
}

/**
 * Full pipeline: mnemonic → .onion address
 */
function mnemonicToOnion(mnemonic) {
    const seed = mnemonicToSeed(mnemonic);
    const keyPair = seedToKeyPair(seed);
    return publicKeyToOnion(keyPair.publicKey);
}

// ─── Tor Key File Generation ────────────────────────────────────────

/**
 * Write Tor hidden service key files from a mnemonic
 * 
 * Creates:
 *   hs_ed25519_secret_key — 96 bytes: header(32) + scalar(32) + righthalf(32)
 *   hs_ed25519_public_key — 64 bytes: header(32) + pubkey(32)
 *   hostname — the .onion address
 * 
 * Returns the .onion address
 */
function writeTorKeys(dir, mnemonic) {
    fs.mkdirSync(dir, { recursive: true });

    const seed = mnemonicToSeed(mnemonic);
    const keyPair = seedToKeyPair(seed);
    const { scalar, righthalf } = expandSecretKey(keyPair.seed);
    const onionAddress = publicKeyToOnion(keyPair.publicKey);

    // ── Secret key file: header(32) + scalar(32) + righthalf(32) = 96 bytes
    const secretHeader = Buffer.alloc(32);
    secretHeader.write('== ed25519v1-secret: type0 ==', 0, 'ascii');
    // Last 3 bytes are already 0x00 from Buffer.alloc
    const secretFile = Buffer.concat([secretHeader, scalar, righthalf]);

    if (secretFile.length !== 96) {
        throw new Error(`Secret key file wrong size: ${secretFile.length} (expected 96)`);
    }
    fs.writeFileSync(path.join(dir, 'hs_ed25519_secret_key'), secretFile);

    // ── Public key file: header(32) + pubkey(32) = 64 bytes
    const publicHeader = Buffer.alloc(32);
    publicHeader.write('== ed25519v1-public: type0 ==', 0, 'ascii');
    const publicFile = Buffer.concat([publicHeader, keyPair.publicKey]);

    if (publicFile.length !== 64) {
        throw new Error(`Public key file wrong size: ${publicFile.length} (expected 64)`);
    }
    fs.writeFileSync(path.join(dir, 'hs_ed25519_public_key'), publicFile);

    // ── Hostname file
    fs.writeFileSync(path.join(dir, 'hostname'), onionAddress + '\n');

    console.log(`[identity] Tor keys written to ${dir}`);
    console.log(`[identity] .onion address: ${onionAddress}`);

    return onionAddress;
}

// ─── Full Identity Object ───────────────────────────────────────────

/**
 * Create a full identity object from a mnemonic
 */
function createIdentity(mnemonic) {
    if (!validateMnemonic(mnemonic)) {
        throw new Error('Invalid BIP-39 mnemonic');
    }
    const seed = mnemonicToSeed(mnemonic);
    const keyPair = seedToKeyPair(seed);
    const onionAddress = publicKeyToOnion(keyPair.publicKey);

    return {
        mnemonic,
        onionAddress,
        publicKey: keyPair.publicKey.toString('hex'),
        words: mnemonic.split(' ')
    };
}

module.exports = {
    generateMnemonic,
    validateMnemonic,
    mnemonicToOnion,
    mnemonicToSeed,
    seedToKeyPair,
    expandSecretKey,
    publicKeyToOnion,
    writeTorKeys,
    createIdentity,
    base32Encode
};

File: octo_app/lib/core/identity.dart
import 'dart:io';
import 'dart:typed_data';
import 'package:bip39/bip39.dart' as bip39;
import 'package:crypto/crypto.dart';
import 'package:cryptography/cryptography.dart';
import 'package:pointycastle/digests/sha3.dart';
import 'package:base32/base32.dart';

class Identity {
  final String mnemonic;
  final String onionAddress;
  final Uint8List publicKey;
  final Uint8List secretKey;

  Identity._({
    required this.mnemonic,
    required this.onionAddress,
    required this.publicKey,
    required this.secretKey,
  });

  /// Generates a new 24-word BIP-39 mnemonic.
  static String generateMnemonic() {
    return bip39.generateMnemonic(strength: 256);
  }

  /// Validates a BIP-39 mnemonic.
  static bool validateMnemonic(String mnemonic) {
    return bip39.validateMnemonic(mnemonic);
  }

  /// Derives the ed25519 keys and Tor v3 onion address from a mnemonic.
  static Future<Identity> fromMnemonic(String mnemonic) async {
    if (!validateMnemonic(mnemonic)) {
      throw Exception('Invalid BIP-39 mnemonic');
    }

    final seed = bip39.mnemonicToSeed(mnemonic);
    final edSeed = seed.sublist(0, 32);

    final hash = sha512.convert(edSeed).bytes;
    final scalar = Uint8List.fromList(hash.sublist(0, 32));
    scalar[0] &= 248;
    scalar[31] &= 127;
    scalar[31] |= 64;

    final rightHalf = Uint8List.fromList(hash.sublist(32, 64));

    final ed25519 = Ed25519();
    final keyPair = await ed25519.newKeyPairFromSeed(edSeed);
    final pubKey = await keyPair.extractPublicKey();
    final pubKeyBytes = Uint8List.fromList(pubKey.bytes);

    // Tor Secret Key format: "== ed25519v1-secret: type0 ==\0\0\0" (32 bytes) + scalar + rightHalf
    final secretHeader = "== ed25519v1-secret: type0 ==\x00\x00\x00".codeUnits;
    final secretKeyBytes = Uint8List.fromList([...secretHeader, ...scalar, ...rightHalf]);

    // Tor Public Key format: "== ed25519v1-public: type0 ==\0\0\0" (32 bytes) + pubkey
    final publicHeader = "== ed25519v1-public: type0 ==\x00\x00\x00".codeUnits;
    final publicKeyBytes = Uint8List.fromList([...publicHeader, ...pubKeyBytes]);

    // Checksum: SHA3_256(".onion checksum" + pubkey + 0x03)
    final prefix = ".onion checksum".codeUnits;
    final version = [0x03];
    final checksumData = Uint8List.fromList([...prefix, ...pubKeyBytes, ...version]);
    final sha3 = SHA3Digest(256);
    final checksum = sha3.process(checksumData).sublist(0, 2);

    final onionData = Uint8List.fromList([...pubKeyBytes, ...checksum, ...version]);
    final onionAddress = base32.encode(onionData).toLowerCase() + '.onion';

    return Identity._(
      mnemonic: mnemonic,
      onionAddress: onionAddress,
      publicKey: publicKeyBytes,
      secretKey: secretKeyBytes,
    );
  }

  /// Writes the Tor hidden service keys to disk.
  Future<void> writeTorKeys(String directoryPath) async {
    final dir = Directory(directoryPath);
    if (!await dir.exists()) {
      await dir.create(recursive: true);
    }

    final secretFile = File('${dir.path}/hs_ed25519_secret_key');
    final publicFile = File('${dir.path}/hs_ed25519_public_key');
    final hostnameFile = File('${dir.path}/hostname');

    await secretFile.writeAsBytes(secretKey);
    await publicFile.writeAsBytes(publicKey);
    await hostnameFile.writeAsString('$onionAddress\n');
  }
}
